[RenderReplaced][Interop] Coerce block sizing for degenerate aspect ratio SVGs.

yulunwu2 opened the pull request:
#### 2a8e0a3b3b2c6ffdd9c12c876b14bfa8ad105977
<pre>
[RenderReplaced][Interop] Coerce block sizing for degenerate aspect ratio SVGs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=297457">https://bugs.webkit.org/show_bug.cgi?id=297457</a>
&lt;<a href="https://rdar.apple.com/156339128">rdar://156339128</a>&gt;

Reviewed by Sammy Gill.

This PR fixes an issue where SVGs with 0px inline intrinsic size (width) were
being ignored in:

RenderReplaced::computeReplacedLogicalWidth()

This is incorrect as 0px is a valid intrinsic width for SVGs and has implications
for sizing the RenderReplaced block. Mishandeling this eventually resulted in
a failure to override the SVG&apos;s block direction intrinsic size (height) to 0px.

Per CSSWG resolution, degenerate aspect ratios derived from SVGs&apos;
width and height (from 0px, or -Npx) should fall back to the viewbox
aspect ratio.
See: <a href="https://github.com/w3c/csswg-drafts/issues/6286#issuecomment-866986544%5C">https://github.com/w3c/csswg-drafts/issues/6286#issuecomment-866986544%5C</a>

Per CSSWG resolution, we should then use the aforementioned aspect ratio
to set the *block* size to match a valid *inline* size.

This means that degenerate aspect ratios resulting from a valid inline
size of 0px should result in a computed block size of 0px.

Note that degenerate aspect ratios from a valid block size of 0px
will instead be computed to respect the viewbox aspect ratio relative to
an explicit inline size.

Note that degenerate aspect ratios from invalid block sizes
(e.g. -1px) will not be respected.
See: <a href="https://github.com/w3c/csswg-drafts/issues/11236#issuecomment-2718502765">https://github.com/w3c/csswg-drafts/issues/11236#issuecomment-2718502765</a>

This PR introduces

shouldRespectZeroIntrinsicWidth()
shouldRespectZeroIntrinsicHeight()

to document this behavior.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/naturalWidth-naturalHeight-width-height.tentative-expected.txt:
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::shouldRespectZeroIntrinsicWidth const):
(WebCore::RenderImage::paint):
* Source/WebCore/rendering/RenderImage.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::shouldRespectZeroIntrinsicWidth const):
(WebCore::RenderReplaced::shouldRespectZeroIntrinsicHeight const):
(WebCore::RenderReplaced::paint):
* Source/WebCore/rendering/RenderReplaced.h:

Canonical link: <a href="https://commits.webkit.org/299085@main">https://commits.webkit.org/299085@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7684a59093bf68f6185a5bf687f4ec6027b7650

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117753 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37431 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28061 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123887 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69767 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c06f4153-152d-4b69-80b5-82c277c898a5) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119631 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38123 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46013 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89369 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/56976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e1c1b88a-7f98-4335-9f78-6eaa8c543616) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120705 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30347 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105580 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69862 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9bbb1fb7-0934-4e41-87f4-8fa916e879d2) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29408 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23695 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67546 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99767 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23875 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126980 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44656 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33613 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98024 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45014 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101805 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97812 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24889 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43194 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21160 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44528 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50202 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43986 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47333 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45677 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/864cf698834cddf480949724714f90a7f23e762a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49457

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ae68a1fe7390a0b65d3c2dce25fcea4b595ff0b6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49457

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f83c16a148fda456bbb440cc6388ccaca9fae15e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49457

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7e08f9ed9f2bf7a80cf567416a0cfd1c8f7aa284. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49457

sammygill commented on Source/WebCore/rendering/RenderImage.cpp:
```diff
@@ -452,6 +453,28 @@ bool RenderImage::shouldDisplayBrokenImageIcon() const
+
+bool RenderImage::hasDefiniteIntrinsicHeight() const
+{
+    if (auto* cachedImage = this->cachedImage()) {
```
You could consider performing an early return if `cahcedImage` is `nullptr` to reduce one level of scoping which could slightly improve readability

yulunwu2 replied:
Do you mean:
if(!this->cachedImage())
    return false;
auto* cachedImage = this->cachedImage();
...

yulunwu2 replied:
...not sure how I missed that. Done and thanks for the review!

sammygill commented on Source/WebCore/rendering/RenderImage.cpp:
```diff
@@ -452,6 +453,28 @@ bool RenderImage::shouldDisplayBrokenImageIcon() const
     return imageResource().errorOccurred();
 }
 
+bool RenderImage::hasDefiniteIntrinsicWidth() const
```
Same suggestions as the one in `hasDefiniteIntrinsicHeight`

yulunwu2 replied:
Done, and thanks for the review!

yulunwu2 commented on Source/WebCore/rendering/RenderReplaced.h:
```diff
@@ -31,6 +31,9 @@ class RenderReplaced : public RenderBox {
 public:
     virtual ~RenderReplaced();
 
+    virtual bool hasDefiniteIntrinsicWidth() const;
+    virtual bool hasDefiniteIntrinsicHeight() const;
```
I'm not a fan of having RenderReplaced() check itself to see if it is a RenderImage() and then look for an SVG as that would make computeReplacedLogicalWidth() harder to read. I agree the function naming could use some work. Renaming this to shouldRespectZeroIntrinsicWidth/Height()

yulunwu2 replied:
After renaming that, I realized a couple things:

1. We don't actually respect zero intrinsic height
2. We should call this in both of:

computeReplacedLogicalWidth
computeReplacedLogicalHeight

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/95995d986b4f83da88c072ba540c49ee31aa27e1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49457

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/66e998e4afcce4b5677318a6af49d09481d57081. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49457

sammygill commented on Source/WebCore/rendering/RenderReplaced.h:
```diff
@@ -31,6 +31,9 @@ class RenderReplaced : public RenderBox {
     virtual ~RenderReplaced();
 
+    virtual bool shouldRespectZeroIntrinsicWidth() const;
+    virtual bool shouldRespectZeroIntrinsicHeight() const;
```
Do we still need this one even though nobody overrides it and will just be false any time it's used?

yulunwu2 replied:
In terms of utility, shouldRespectZeroIntrinsicHeight() isn't really doing anything. However, it is documenting that zero intrinsic height is not being respected so I'm partial to keeping it. Happy to remove it if you feel strongly about this.

sammygill replied:
I think in its current state it's a little confusing if you were to stumble upon this and figure out what's going on. 

The function declaration is `virtual` which when combined with the name of the function implies that there should be some situations in which we would respect zero intrinsic height and in some situations we would not, but that is not the case given the only implementation on `RenderReplaced`. That somewhat begs the question of whether there are any scenarios in which we should respect zero intrinsic height and it's not really clear from the code at least. The commit message mentions using a 0px natural height but I'm not sure if that answers the question since I don't think the code currently can detect a natural height of 0.

IMO this could be best explained as a comment alongside `shouldRespectZeroIntrinsicWidth`, but I'll leave it up to you. We should at minimum probably not make this virtual to help reduce the confusion and not add a vpointer for it. (Note that there's also a chance that someone could come through in a "cleanup," patch and remove it since it always returns the same value)

yulunwu2 committed:
[RenderReplaced][Interop] Coerce block sizing for degenerate aspect ratio SVGs.
https://bugs.webkit.org/show_bug.cgi?id=297457
<rdar://156339128>

Reviewed by Sammy Gill.

This PR fixes an issue where SVGs with 0px inline intrinsic size (width) were
being ignored in:

RenderReplaced::computeReplacedLogicalWidth()

This is incorrect as 0px is a valid intrinsic width for SVGs and has implications
for sizing the RenderReplaced block. Mishandeling this eventually resulted in
a failure to override the SVG's block direction intrinsic size (height) to 0px.

Per CSSWG resolution, degenerate aspect ratios derived from SVGs'
width and height (from 0px, or -Npx) should fall back to the viewbox
aspect ratio.
See: https://github.com/w3c/csswg-drafts/issues/6286#issuecomment-866986544%5C

Per CSSWG resolution, we should then use the aforementioned aspect ratio
to set the *block* size to match a valid *inline* size.

This means that degenerate aspect ratios resulting from a valid inline
size of 0px should result in a computed block size of 0px.

Note that degenerate aspect ratios from a valid block size of 0px
will instead be computed to respect the viewbox aspect ratio relative to
an explicit inline size.

Note that degenerate aspect ratios from invalid block sizes
(e.g. -1px) will not be respected.
See: https://github.com/w3c/csswg-drafts/issues/11236#issuecomment-2718502765

This PR introduces

shouldRespectZeroIntrinsicWidth()
shouldRespectZeroIntrinsicHeight()

to document this behavior.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-img-element/naturalWidth-naturalHeight-width-height.tentative-expected.txt:
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::shouldRespectZeroIntrinsicWidth const):
(WebCore::RenderImage::paint):
* Source/WebCore/rendering/RenderImage.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::shouldRespectZeroIntrinsicWidth const):
(WebCore::RenderReplaced::shouldRespectZeroIntrinsicHeight const):
(WebCore::RenderReplaced::paint):
* Source/WebCore/rendering/RenderReplaced.h:

Canonical link: https://commits.webkit.org/299085@main
